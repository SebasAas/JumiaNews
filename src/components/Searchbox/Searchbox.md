## Importing Component

```js static
import Searchbox from '<RootFolder>';
```

<hr/>

## Basic Searchbox:

Return Searchbox

### Default

```html
<Searchbox />
```

<hr/>

## Searchbox onSubmit function to close hamburger menu

### Expected

- Function used to close the hamburger menu on submit search (mobile)

### Default

- null

### Return

- false

```html
<Searchbox setOpen={fn()} />
```
