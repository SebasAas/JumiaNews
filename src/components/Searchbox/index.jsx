import React, { useState, useEffect } from 'react'
import { Link, useNavigate } from "react-router-dom";

// Styles
import styled from 'styled-components';
import { primaryOrange } from '../../theme/colors.js'

// Icons
import { SearchIcon } from '../../assets/Icons';

function Searchbox({ setOpen = null }) {
  const navigate = useNavigate();
  const [termSearch, setTermSearch] = useState('');
  const [colorFill, setColorFill] = useState('');

  const handlePropagationColor = (type) => {
    if (type === 'focus') {
      setColorFill(primaryOrange)
    } else if (type === 'hover') {
      setColorFill('#6d6d6d')
    } else {
      setColorFill('#969696')
    }
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    if (setOpen !== null) {
      setOpen(false)
    }
    navigate(`/section/${termSearch}`);
  }

  return (
    <form onSubmit={(e) => handleSubmit(e)} className='relative'>
      <Input
        className='input-search'
        data-testid="input-search"
        type="text"
        placeholder="search"
        onFocus={() => handlePropagationColor('focus')}
        onMouseOver={() => handlePropagationColor('hover')}
        onBlur={() => handlePropagationColor('leave')}
        value={termSearch}
        onChange={(e) => setTermSearch(e.target.value)}
        style={setOpen && { color: 'white' }}
      />
      <ButtonBox onClick={(e) => handleSubmit(e)} colorFill={colorFill}>
        <Link to={`/section/${termSearch}`}>
          {SearchIcon()}
        </Link>
      </ButtonBox>
    </form>
  )
}

const Input = styled.input`
  width: 100%;
  border: 2px solid #969696;
  border-radius: 50px;
  padding: 12px 30px;
  background: transparent;

  &:hover {
    border-color: #808080;
  }

  &:focus {
    outline: none !important; 
    border-color: ${primaryOrange};
  }
`

const ButtonBox = styled.button`
  background: transparent;
  position: absolute;
  color: ${({ colorFill }) => colorFill ? colorFill : '#969696'};
  top: 5px;
  right: 5px;
`



export default Searchbox
