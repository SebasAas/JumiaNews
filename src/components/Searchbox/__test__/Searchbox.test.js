import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom'

import 'jest-styled-components'

// Components
import Searchbox from '../index'

describe('Rendering Searchbox Component', () => {

  afterEach(() => {
    jest.resetAllMocks();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  test('should change the value', () => {
    const termToSearch = 'New York Times'

    const { queryByPlaceholderText, getByRole } = render(<Searchbox />, { wrapper: MemoryRouter })
    const searchInput = queryByPlaceholderText('search')

    fireEvent.change(searchInput, { target: { value: encodeURI(termToSearch) } })

    expect(getByRole('link')).toHaveAttribute('href', encodeURI(`/section/${termToSearch}`))
  })

  test('should assign the value of the search as id in the route', () => {
    const termToSearch = 'BBC International'

    const { queryByPlaceholderText } = render(<Searchbox />, { wrapper: MemoryRouter })
    const searchInput = queryByPlaceholderText('search')

    fireEvent.change(searchInput, { target: { value: encodeURI(termToSearch) } })

    expect(searchInput.value).toBe(encodeURI(termToSearch))
  })

  test('should render input with variation', async () => {
    const { queryByPlaceholderText } = render(<Searchbox />, { wrapper: MemoryRouter })
    const searchInput = queryByPlaceholderText('search')

    expect(searchInput).toMatchSnapshot()
  })

})