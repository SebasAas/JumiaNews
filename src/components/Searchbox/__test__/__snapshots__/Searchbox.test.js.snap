// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Searchbox Component should render input with variation 1`] = `
.c0 {
  width: 100%;
  border: 2px solid #969696;
  border-radius: 50px;
  padding: 12px 30px;
  background: transparent;
}

.c0:hover {
  border-color: #808080;
}

.c0:focus {
  outline: none !important;
  border-color: #FF7B51;
}

<input
  class="c0 input-search"
  data-testid="input-search"
  placeholder="search"
  type="text"
  value=""
/>
`;
